### DISCUSSION TITLE
10 Second Countdown Lose Condition

### DISCUSSION THREAD

**Discussion by christopherpeers** [cyberbotics/wrestling#33] [General]
Howdy,

To my understanding, the 10 second countdown for each player decreases at the exact same time starting from when the first player falls. Therefore, if both players fall within a second of each other, the number displayed (and counted) for each player will be exactly the same.

Subsequently, if both players remain on the floor until the timer decreases to zero, you would assume this was a draw or there would be a millisecond time difference as the reason for one side winning, or simply the side with the higher score at the time wins.

After reviewing the replay for Chrake Peith vs Pemile, in run_trigger #1087, I found out that when this situation occurs, it is red (the attacking player) that is declared the winner. This is because the referee code assigns the knockout to a single parameter "ko" and since blue is always checked after red, when the referee comes to check who is KO, it is always blue.

Lines 99-100

if self.ko_count[i] > 10000:  # 10 seconds
            ko = i

Lines 112-117

if ko == 0:
            print('Red is KO. Blue wins!')
            performance = 0
elif ko == 1:
            print('Blue is KO. Red wins!')
            performance = 1

It is possible to check this yourself by starting a match with both robots on the floor, and you will find red will always win.

I hope highlighting this will do some good for the competition, even though we are still in "walk around for the highest score and then hope for the best" purgatory.

**Comment by omichel** [comment]
Thank you for reporting this.

> Therefore, if both players fall within a second of each other, the number displayed (and counted) for each player will be exactly the same.

This is not exactly true as [the test is performed every 200 ms](https://github.com/cyberbotics/wrestling/blob/main/controllers/referee/referee.py#L71). But basically, what you discovered is correct even thought is a matter of 200 ms and not one second, and it is probably a flaw that we should address.

From there, I see two minor changes to improve the rules:

1. We decrease the test frequency to 20 ms instead of 200 ms, which will make that if a robot fell more than 20 ms after its opponent it will win whatever its color. Note: we cannot go below those 20 ms as this is the basic time step defined in the world file.
2. In case of perfect equality, e.g., when the two robots fall at the very same time (within the 20 ms resolution), then, the robot with the highest coverage wins.

Would this address the problem?

If nobody disagrees with this minor rule change, we are going to implement it very soon.

**Reply by omichel** [reply]
Please see #35 which implements this minor rule change.

**Reply by zicklag** [reply]
The new rule sounds good to me. 👍

**Reply by omichel** [reply]
All right, this was merged right now and should be in production from now on.
